docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,649 @@
+that match them. The downside is that we are adding a ref to the initial
+advertised refs, each of theses refs is around 250 bytes. If we consider a popular
+repository such as https://github.com/facebook/react/ you can see that they have
+nearly 10,000 issues (including both Open and Closed, which we must). That would 
+mean that the initial ref advertisment in any replication of this repository
+would be ~2.5Mb. If we could use v2 of the git pack protocol this problem would
+go away but alas, we must use v1. 
+
+To get around the problem we can add a parameter to the radicle URL which
+indicates either a single object ID or a type name which we wish to fetch
+from the remote, this would then be passed as a custom header to the transport
+and interpreted on the server. In this manner we can allow clients to choose
+when they want to replicate collaborative objects, which would allow staged
+fetches where we first fetch the repository and identities, and then fetch 
+collaborative objects.
+
+Therefore we add two parameters to a radicle URN:
+
+- `collab_object_id`: specifies that a server should only consider references to the given object ID
+- `collab_type_name`: specifies that a server should only consider references to the given type name
+
+These parameters are emitted by the radicle p2p transport as headers in a 
+similar fashion to the `n=` parameter for the nonce or the `ls=true`
+parameter for selecting the `upload-pack-ls` service. The Radicle git server
+can then use these parameters to filter the refs it operates on.
+
+
+### Updating objects
+
+To make a change to an object we load the existing messages for an object. The 
+application developer provides us with the binary representation of the change
+to that object. We apply the change and ensure that the new object state still
+matches the object schema. At this point the state of the object may depend on
+many contributions from the tracking graph - not just the ones in our own view
+of the project. We now create a commit with our new change in it, referencing
+all the commits containing the direct dependencies of the change as parents.
+
+
+### Schema extensions
+
+#### `rad_signed_by`
+
+Many collaborative data structures will need to make statements about who is
+allowed to change what parts of a structure. To achieve this we extend the 
+json schema language with some custom metadata, the `rad_signed_by` property. This
+property can be placed on any `object` schema. It's value is an
+object with two keys, an array of properties which must be signed, and array of 
+radicle URNs who's signature must be present.
+
+This property implies a required `signatures` property which contains a mapping
+from URN to a description of the signature. (Maybe a JWS?).
+
+Any schema which has this property will result in some additional validation.
+Librad will encode the relevant keys of the target object using canonical JSON
```

Comment:
> canonical JSON

Which implies that objects can't have floating-point values 🙃

Reply:
Ugh, good point, I've modified this to specify CBOR instead.

Reply:
Mya, not a [silver bullet](https://datatracker.ietf.org/doc/html/rfc7049#section-3.9). Canonicalisation of floats specifically is only defined non-normatively.

Reply:
Ugh, any other suggestions?

Reply:
It’s not entirely clear to me what the actual payload for the signature is — when it’s an update, would it be only the actual automerge change, or the result of applying the change? If automerge supports difficult-to-canonicalise types such as floats, how does it handle them? If a change is produced deterministically (ie. if I put a float field as 0.23417, something will prevent me to accidentally create the same change but with a rounding error), and changes have an identity, does canonicalisation actually matter?

Reply:
The payload is the given properties of the realization of the state of the automerge document after having applied the automerge change. i.e the result of applying the change. Automerge has both 32 and 64 floating point numbers which are little endian encoded in the document. The automerge libraries are designed so that these types round trip through serialization and deserialization, so in this sense they are deterministic. We could stipulate that we use canonical CBOR with the float type being the type of the underlying automerge value.

The intention of the `rad_signed_by` schema is to impose constraints on parts of the state of the CRDT, so we can't use the change  directly. For example, a user might update their comment by inserting a character, the underlying op log would have two operations: one inserting a character and one updating the siganature, but we want to validate that the signature is valid over the author and the whole comment, so we can't use the change itself as the item over which to validate the signature.

Reply:
Interesting. I suppose it’s potentially useful to retain the signatures in the document, so they can be validated on the frontend as well. The lossy typing through JSON requires to define canonicalisation rules, though, which interestingly are the opposite of what RFC 7049 says: we need to use the _largest_ representation.

Aside: it’s an interesting choice that automerge doesn’t seem to concern itself with Unicode normalisation, given that applications typically won’t display diffs which would show that U+212B, U+0041U+030A and U+00C5 are considered three different characters. This means that we shouldn’t normalise either, because that would increase the number of possible inputs to the signature function.

So, my take would be:

—-
To compute a signature, the values at the keys specified in `rad_signed_by.keys` are looked up from the document in the order given, and converted back-to-back into CBOR according to the following canonicalisation rules:

* `integer` values are encoded as major types 0 and 1, respectively, with additional value 27 (64-bit) 
* `number` types are encoded as major type 7 with 5-bit value 27 (IEEE 754 double-precision float)
* `object` types are encoded as major type 5 (maps), applying the canonicalisation rules as per RFC 7049, section 2.3
* The remaining schema types have unambiguous mappings to CBOR types


When the input is in the automerge data model, the mappings are as follows:

* f32,f64: major type 7,27
* int: major type 0 or 1,27
* uint: major type 1,27
* bytes: major type 2, where the length is represented as a 64-bit unsigned integer (major type 1,27)
* str: major type 3, where the length is represented as a 64-bit unsigned integer (major type 1,27)
* boolean: major type 7, 20 or 21 respectively 
* counter, timestamp: as their corresponding integer value
* null: major type 7, 22


Cursors, undefined values, and constructions which can not be represented as definite-length arrays / maps in CBOR are illegal for signed values.

Finally, the signature is computed over the binary CBOR representation and stored in the `signature` property. To verify the signature, the payload is computed in the same way, and verified against the `signature` and `key` properties.